src/api/utils/tradeQuote/tradequote.ts

```diff
@@ -0,0 +1,404 @@
+
+const SCALE = BigNumber.from(10).pow(18);
+
+/**
+ * @title TradeQuoteAPI
+ * @author Set Protocol
+ *
+ * A utility library to generate trade quotes for token pairs associated with a
+ * set for Ethereum and Polygon chains. Uses 0xAPI to get quote and requires a valid
+ * 0x api key.
+ */
+
+export class TradeQuoteAPI {
+  private setToken: SetTokenAPI;
+  private tokenMap: CoinGeckoTokenMap;
+  private largeTradeGasCostBase: number = 150000;
+  private tradeQuoteGasBuffer: number = 5;
+  private zeroExApiKey: string;
+
+  constructor(setToken: SetTokenAPI, zeroExApiKey: string = '') {
+    this.setToken = setToken;
+    this.zeroExApiKey = zeroExApiKey;
+  }
+
+  /**
+   * Generates a trade quote for a token pair in a SetToken. This method requires
+   * a token metadata map (passed with the options) which can be generated using
+   * the CoinGeckoDataService in '.api/utils/coingecko.ts'.
+   *
+   * @param  options  QuoteOptions: options / config to generate the quote
+   * @return          TradeQuote: trade quote object
+   */
+  async generate(options: QuoteOptions): Promise<TradeQuote> {
+    this.tokenMap = options.tokenMap;
+    const feePercentage = options.feePercentage || 0;
+    const isFirmQuote = options.isFirmQuote || false;
+    const chainId = options.chainId;
+    const exchangeAdapterName = ZERO_EX_ADAPTER_NAME;
+
+    const {
+      fromTokenAddress,
+      toTokenAddress,
+      fromAddress,
+    } = this.sanitizeAddress(options.fromToken, options.toToken, options.fromAddress);
+
+    const amount = this.sanitizeAmount(fromTokenAddress, options.rawAmount);
+
+    const setOnChainDetails = await this.setToken.fetchSetDetailsAsync(
+      fromAddress, [fromTokenAddress, toTokenAddress]
+    );
+
+    const fromTokenRequestAmount = this.calculateFromTokenAmount(
+      setOnChainDetails,
+      fromTokenAddress,
+      amount
+    );
+
+    const {
+      fromTokenAmount,
+      fromUnits,
+      toTokenAmount,
+      toUnits,
+      calldata,
+      zeroExGas,
+    } = await this.fetchZeroExQuote( // fetchQuote (and switch...)
+      fromTokenAddress,
+      toTokenAddress,
+      fromTokenRequestAmount,
+      setOnChainDetails.manager,
+      (setOnChainDetails as any).totalSupply, // Typings incorrect,
+      chainId,
+      isFirmQuote,
+      options.slippagePercentage
+    );
+
+    // Sanity check response from quote APIs
+    this.validateQuoteValues(
+      setOnChainDetails,
+      fromTokenAddress,
+      toTokenAddress,
+      fromUnits,
+      toUnits
+    );
+
+    const gas = this.estimateGasCost(zeroExGas);
+
+    const coinGecko = new CoinGeckoDataService(chainId);
+    const coinPrices = await coinGecko.fetchCoinPrices({
+      contractAddresses: [this.chainCurrencyAddress(chainId), fromTokenAddress, toTokenAddress],
+      vsCurrencies: [ USD_CURRENCY_CODE, USD_CURRENCY_CODE, USD_CURRENCY_CODE ],
+    });
+
+    const gasOracle = new GasOracleService(chainId);
```

Comment:
Same with the coin gecko bit, this is a terrible dependency that can disrupt the flow. Want to abstract, add defaults, and make fault tolerant in a follow up!

Reply:
Now tracking in #64